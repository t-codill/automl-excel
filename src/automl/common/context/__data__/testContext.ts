import { PageNames } from "../../PageNames";
import { Logger } from "../../utils/logger";
import { IWorkspaceContextProps, WorkspaceError } from "../IWorkspaceProps";
import { WorkspaceFlight } from "../WorkspaceFlight";

export function getLogCustomEventSpy(): jest.SpyInstance<ReturnType<Logger["logCustomEvent"]>> {
    const spy = jest.spyOn(Logger.prototype, "logCustomEvent");
    spy.mockImplementation(jest.fn());
    return spy;
}
let pageName: PageNames = PageNames.Unknown;
export const testContext: IWorkspaceContextProps = {
    get pageName(): PageNames { return pageName; },
    get theme(): string { return "light"; },
    discoverUrls: {},
    flight: new WorkspaceFlight(""),
    location: "eastus",
    resourceGroupName: "testResource",
    subscriptionId: "00000000-0000-0000-0000-000000000000",
    logger: new Logger("test"),
    getToken: () => "testToken",
    setPageName: (value: PageNames) => { pageName = value; },
    workspaceName: "testWorkSpace",
    setNavigationBarButtons(): void { return; },
    onError: (err: WorkspaceError) => { throw err; },
    clearErrors: () => { return; },
    setHeader: () => { return; },
    setLoading: () => { return; }
};
