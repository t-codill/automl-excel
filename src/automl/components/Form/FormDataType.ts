import { BasicTypes } from "../../common/BasicTypes";

export type FormDataType = BasicTypes | string[];
