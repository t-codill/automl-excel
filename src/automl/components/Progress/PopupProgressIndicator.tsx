import { Dialog, DialogType, ProgressIndicator } from "office-ui-fabric-react";
import * as React from "react";

interface IPopupProgressIndicatorProps {
    title?: string;
    description?: string;
    percentComplete?: number;
    onDismiss?(): void;
}

export class PopupProgressIndicator
    extends React.PureComponent<IPopupProgressIndicatorProps> {
    public render(): React.ReactNode {

        return <Dialog
            hidden={false}
            minWidth={800}
            onDismiss={this.props.onDismiss}
            modalProps={{
                isBlocking: !this.props.onDismiss
            }}
            dialogContentProps={{
                type: DialogType.normal,
                title: this.props.title
            }}
        >
            <ProgressIndicator
                description={this.props.description}
                percentComplete={this.props.percentComplete} />
        </Dialog>;
    }
}
