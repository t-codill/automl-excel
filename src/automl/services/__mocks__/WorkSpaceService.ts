import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { IVMSizeResult } from "../WorkSpaceService";

export class WorkSpaceService {

    private readonly sampleWorkSpace: AzureMachineLearningWorkspacesModels.Workspace = {
        id: "Test_ID",
        name: "Test_WorkSpace",
        location: "eastus",
        discoveryUrl: "https://dummyurl.dummy",
    };

    private readonly sampleCompute: AzureMachineLearningWorkspacesModels.ComputeResource = {
        id: "Test_ID",
        name: "Test_Compute",
        location: "eastus",
    };

    private readonly sampleVMSize: AzureMachineLearningWorkspacesModels.VirtualMachineSize = {
        name: "Standard_DS1_v2",
        family: "standardDSv2Family",
        vCPUs: 1,
        osVhdSizeMB: 1047552,
        maxResourceVolumeMB: 7168,
        memoryGB: 3.5,
        lowPriorityCapable: true,
        premiumIO: true
    };

    public async listWorkspaces(): Promise<AzureMachineLearningWorkspacesModels.Workspace[] | undefined> {
        return [this.sampleWorkSpace];
    }

    public async tryGetWorkspace(): Promise<AzureMachineLearningWorkspacesModels.Workspace | null | undefined> {
        return this.sampleWorkSpace;
    }

    public async getWorkspace(): Promise<AzureMachineLearningWorkspacesModels.Workspace | undefined> {
        return this.sampleWorkSpace;
    }

    public async tryGetCompute(): Promise<AzureMachineLearningWorkspacesModels.ComputeResource | null | undefined> {
        return this.sampleCompute;
    }

    public async listComputes(): Promise<AzureMachineLearningWorkspacesModels.ComputeResource[] | undefined> {
        return [this.sampleCompute];
    }

    public async createCompute(): Promise<AzureMachineLearningWorkspacesModels.ComputeResource | undefined> {
        return this.sampleCompute;
    }

    public async listVmSizes(): Promise<IVMSizeResult | undefined> {
        return { ...[this.sampleVMSize], amlCompute: [this.sampleVMSize] };
    }

    public dispose(): void {
        return;
    }

    public reset(): void {
        return;
    }
}
