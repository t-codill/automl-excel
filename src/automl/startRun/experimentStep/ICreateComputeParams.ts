export interface ICreateComputeParams {
    name: string;
    vmSize: string;
    minNodeCount: string;
    maxNodeCount: string;
}
