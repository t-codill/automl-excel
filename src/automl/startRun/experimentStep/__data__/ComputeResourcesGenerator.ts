import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";

export type ComputeType = "Compute" | "AKS" | "AmlCompute" | "VirtualMachine" | "HDInsight" | "DataFactory" | "Databricks" | "DataLakeAnalytics";

export const getComputeResources = (
    testComputeType: ComputeType,
    testProvisioningState: AzureMachineLearningWorkspacesModels.ProvisioningState | undefined) => {
    let compute: AzureMachineLearningWorkspacesModels.ComputeResource[];
    if (testComputeType === "AmlCompute") {
        compute = [
            {
                name: "Compute1",
                id: "ComputeId1",
                properties: {
                    computeType: testComputeType,
                    provisioningState: testProvisioningState,
                    properties: {
                        scaleSettings: {
                            maxNodeCount: 2,
                            minNodeCount: 0
                        }
                    }
                }
            },
            {
                name: "Compute2",
                id: "ComputeId2",
                properties: {
                    computeType: testComputeType,
                    provisioningState: testProvisioningState,
                    properties: {
                        scaleSettings: {
                            maxNodeCount: 2,
                            minNodeCount: 1
                        }
                    }
                }
            },
            {
                name: "Compute No ID",
                properties: {
                    computeType: testComputeType,
                    provisioningState: testProvisioningState,
                    properties: {
                        scaleSettings: {
                            maxNodeCount: 2,
                            minNodeCount: 1
                        }
                    }
                }
            },
            {
                name: "Compute unusable",
                properties: {
                    computeType: testComputeType,
                    provisioningState: testProvisioningState,
                    properties: {
                        scaleSettings: {
                            maxNodeCount: 2,
                            minNodeCount: 1
                        },
                        nodeStateCounts: {
                            unusableNodeCount: 2
                        }
                    }
                }
            },
            {
                name: "Compute No Property",
                id: "ComputeId2"
            }
        ];
    }
    else {
        compute = [
            {
                name: "Compute1",
                id: "ComputeId1",
                properties: {
                    // tslint:disable-next-line: no-any
                    computeType: testComputeType as any,
                    provisioningState: testProvisioningState
                }
            }
        ];
    }
    return compute;
};
