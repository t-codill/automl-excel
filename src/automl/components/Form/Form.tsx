import * as React from "react";
import { FormBaseInput } from "./FormBaseInput";
import { FormContext } from "./FormContext";
import { FormDataType } from "./FormDataType";

export interface IFormProps<TData extends { [key in keyof TData]: FormDataType }> {
    onSubmit(value: TData): void;
    onUpdated?(key: keyof TData, value: TData[keyof TData]): void;
}
export interface IFormStates<TData extends { [key in keyof TData]: FormDataType }> {
    onSubmit(value: TData): void;
}

export class Form<TData extends { [key in keyof TData]: FormDataType }> extends React.Component<IFormProps<TData>, IFormStates<TData>> {
    private readonly _inputs: Set<FormBaseInput<TData, keyof TData, TData[keyof TData]>> = new Set();
    public render(): React.ReactNode {
        return (
            <FormContext.Provider value={{
                mount: this._mount,
                unmount: this._unmount,
                onUpdated: this.props.onUpdated
            }}>
                <form
                    onSubmit={this._onSubmit}
                    noValidate={true}
                    autoComplete="off">
                    {this.props.children}
                </form>
            </FormContext.Provider>
        );

    }
    private readonly _mount = (input: FormBaseInput<TData, keyof TData, TData[keyof TData]>) => {
        this._inputs.add(input);
        return;
    }
    private readonly _unmount = (input: FormBaseInput<TData, keyof TData, TData[keyof TData]>) => {
        this._inputs.delete(input);
        return;
    }
    private readonly _onSubmit = (event: React.FormEvent<HTMLElement>): void => {
        event.preventDefault();
        const val: TData = Object.create(null);
        let valid = true;
        this._inputs.forEach((input) => {
            val[input.props.field] = input.getValue();
            valid = !input.validate() && valid;
        });
        if (valid) {
            this.props.onSubmit(val);
        }
    }
}
