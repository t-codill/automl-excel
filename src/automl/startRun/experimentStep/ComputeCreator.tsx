import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { DefaultButton, Panel, PanelType, PrimaryButton } from "office-ui-fabric-react";
import * as React from "react";
import { Accordion } from "../../components/Accordion/Accordion";
import { BaseComponent } from "../../components/Base/BaseComponent";
import { Form } from "../../components/Form/Form";
import { FormTextInput } from "../../components/Form/FormTextInput";
import { Validators } from "../../components/Form/Validators";
import { PopupProgressIndicator } from "../../components/Progress/PopupProgressIndicator";
import { IVMSizeResult, WorkSpaceService } from "../../services/WorkSpaceService";
import { ICreateComputeParams } from "./ICreateComputeParams";
import { MaxNodeInput } from "./MaxNodeInput";
import { MinNodeInput } from "./MinNodeInput";
import { VmSizeSelector } from "./VmSizeSelector";

export interface IComputeCreatorProps {
    vmSizes: IVMSizeResult | undefined;
    onComputeCreated(compute: AzureMachineLearningWorkspacesModels.ComputeResource): void;
    onCancel(): void;
}

export interface IComputeCreatorState {
    creating: boolean;
    minNodeCount: string;
}

export class ComputeCreator extends BaseComponent<IComputeCreatorProps, IComputeCreatorState,
    {
        workSpaceService: WorkSpaceService;
    }> {
    protected readonly serviceConstructors = {
        workSpaceService: WorkSpaceService
    };

    protected readonly getData = undefined;

    constructor(props: IComputeCreatorProps) {
        super(props);
        this.state = {
            creating: false,
            minNodeCount: "0"
        };
    }

    public render(): React.ReactNode {
        return <>
            {!this.state.creating &&
                <Panel
                    isOpen={true}
                    type={PanelType.medium}
                    onDismiss={this.cancel}
                    isHiddenOnDismiss={true}
                    headerText="Create a New Compute"
                    closeButtonAriaLabel="Close"
                >
                    <Form onSubmit={this.createCompute} onUpdated={this.onFormUpdated}>
                        <FormTextInput<ICreateComputeParams, "name">
                            field="name"
                            required={true}
                            label="Compute name"
                            placeholder="Compute name"
                            ariaLabel="Please enter compute name here"
                            validators={[
                                Validators.required("Compute name is required"),
                                Validators.regex(/^[a-zA-Z][a-zA-Z0-9-]{1,15}$/,
                                    "Compute name should start with a letter, be between 2 and 16 character, and only include letters (a-zA-Z), numbers (0-9) and '-'")]}
                        />
                        <VmSizeSelector vmSizes={this.props.vmSizes} />
                        <Accordion
                            exclusive={true}
                            items={[
                                {
                                    title: "Additional Settings",
                                    collapsed: true,
                                    disabled: false,
                                    element: <>
                                        <MinNodeInput />
                                        <MaxNodeInput minNodeCount={this.state.minNodeCount} />
                                    </>
                                }
                            ]} />
                        <div className="form-footer">
                            <DefaultButton text="Cancel" onClick={this.cancel} />
                            <PrimaryButton text="Create" type="submit" />
                        </div>
                    </Form>
                </Panel>
            }
            {this.state.creating &&
                <PopupProgressIndicator
                    title="Creating a new compute..."
                    description="Creating a new compute. This may take a couple minutes."
                />}
        </>;
    }

    private readonly onFormUpdated = (key: keyof ICreateComputeParams, value: ICreateComputeParams[keyof ICreateComputeParams]) => {
        if (!value) {
            return;
        }
        if (key === "minNodeCount") {
            this.setState({ minNodeCount: value });
        }
    }

    private readonly cancel = () => {
        this.props.onCancel();
    }

    private readonly createCompute = async (data: ICreateComputeParams) => {
        if (!data
            || !data.name
            || !data.vmSize
            || !data.minNodeCount
            || !data.maxNodeCount) {
            return;
        }

        this.setState({
            creating: true
        });
        this.logUserAction("CreateCompute", { data: JSON.stringify(data) });
        const compute = await this.services.workSpaceService.createCompute(
            data.name,
            data.vmSize,
            parseInt(data.minNodeCount, 10),
            parseInt(data.maxNodeCount, 10)
        );

        this.setState({ creating: false });

        if (!compute) {
            return;
        }

        this.props.onComputeCreated(compute);
    }
}
