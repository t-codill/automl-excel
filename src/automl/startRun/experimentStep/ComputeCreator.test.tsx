import { shallow, ShallowWrapper } from "enzyme";
import { Dictionary } from "lodash";
import { DefaultButton } from "office-ui-fabric-react";
import * as React from "react";
import { getLogCustomEventSpy, testContext } from "../../common/context/__data__/testContext";
import { Form } from "../../components/Form/Form";
import { FormDataType } from "../../components/Form/FormDataType";
import { PopupProgressIndicator } from "../../components/Progress/PopupProgressIndicator";
import { WorkSpaceService } from "../../services/WorkSpaceService";
import { ComputeCreator, IComputeCreatorProps, IComputeCreatorState } from "./ComputeCreator";
import { ICreateComputeParams } from "./ICreateComputeParams";

jest.mock("../../services/WorkSpaceService");

describe("ComputeCreator", () => {
    let root: ShallowWrapper<IComputeCreatorProps, IComputeCreatorState, ComputeCreator>;
    let onCancel: jest.Mock;
    let onComputeCreated: jest.Mock;
    let createComputeSpy: jest.SpyInstance<ReturnType<WorkSpaceService["createCompute"]>>;
    beforeEach(() => {
        onCancel = jest.fn();
        onComputeCreated = jest.fn();
        createComputeSpy = jest.spyOn(WorkSpaceService.prototype, "createCompute");
        root = shallow(<ComputeCreator
            onCancel={onCancel}
            onComputeCreated={onComputeCreated}
            vmSizes={{
                amlCompute: []
            }}
        />);
    });
    it("should render", () => {
        expect(root)
            .toMatchSnapshot();
    });
    it("should show loading icon", () => {
        root.setState({ creating: true });
        expect(root.exists(PopupProgressIndicator))
            .toBe(true);
    });
    it("should cancel", () => {
        root.find(DefaultButton)
            .simulate("click");
        expect(onCancel)
            .toBeCalled();
    });
    it("should cancel", () => {
        root.find(DefaultButton)
            .simulate("click");
        expect(onCancel)
            .toBeCalled();
    });
    describe("onFormUpdated", () => {
        let onFormUpdated: (key: keyof ICreateComputeParams, value: ICreateComputeParams[keyof ICreateComputeParams]) => void;
        beforeEach(() => {
            const formProps = root
                .find(Form)
                .props();
            if (formProps.onUpdated) {
                onFormUpdated = formProps.onUpdated;
            }
        });
        it("should update minNodeCount", () => {
            onFormUpdated("minNodeCount", "12");
            expect(root.state("minNodeCount"))
                .toBe("12");
        });
        it("should not update without value", () => {
            onFormUpdated("minNodeCount", "");
            expect(root.state("minNodeCount"))
                .toBe("0");
        });
        it("should not update for other props", () => {
            onFormUpdated("maxNodeCount", "12");
            expect(root.state("minNodeCount"))
                .toBe("0");
        });
    });

    describe("create", () => {
        let onSubmit: (data: ICreateComputeParams & Dictionary<FormDataType>) => void;
        let logSpy: jest.SpyInstance;
        beforeEach(() => {
            const formProps = root
                .find(Form)
                .props();
            if (formProps.onSubmit) {
                onSubmit = formProps.onSubmit;
            }
            logSpy = getLogCustomEventSpy();
        });
        it("should not submit if data missing", async () => {
            onSubmit({
                maxNodeCount: "",
                minNodeCount: "",
                name: "",
                vmSize: ""
            });
            expect(root.state("creating"))
                .toBe(false);
            await Promise.resolve();
            expect(createComputeSpy)
                .not
                .toBeCalled();
        });
        it("should creat", async () => {
            const data = {
                maxNodeCount: "2",
                minNodeCount: "1",
                name: "test",
                vmSize: "testSize"
            };
            onSubmit(data);
            expect(root.state("creating"))
                .toBe(true);
            await Promise.resolve();
            expect(logSpy)
                .toBeCalledWith("_CreateCompute_UserAction", testContext, expect.objectContaining({ component: "CreateCompute", data: JSON.stringify(data) }));
            expect(createComputeSpy)
                .toBeCalledWith(data.name,
                    data.vmSize,
                    parseInt(data.minNodeCount, 10),
                    parseInt(data.maxNodeCount, 10));
            expect(root.state("creating"))
                .toBe(false);
            expect(onComputeCreated)
                .toBeCalled();
        });
        it("should not callback if create compute return undefined", async () => {
            createComputeSpy.mockReturnValueOnce(Promise.resolve(undefined));
            const data = {
                maxNodeCount: "2",
                minNodeCount: "1",
                name: "test",
                vmSize: "testSize"
            };
            onSubmit(data);
            expect(root.state("creating"))
                .toBe(true);
            await Promise.resolve();
            expect(onComputeCreated)
                .not
                .toBeCalled();
        });
    });
});
