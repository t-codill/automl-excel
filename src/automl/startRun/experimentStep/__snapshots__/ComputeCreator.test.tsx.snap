// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComputeCreator should render 1`] = `
<Fragment>
  <StyledPanelBase
    closeButtonAriaLabel="Close"
    headerText="Create a New Compute"
    isHiddenOnDismiss={true}
    isOpen={true}
    onDismiss={[Function]}
    type={3}
  >
    <Form
      onSubmit={[Function]}
      onUpdated={[Function]}
    >
      <FormTextInput
        ariaLabel="Please enter compute name here"
        field="name"
        label="Compute name"
        placeholder="Compute name"
        required={true}
        validators={
          Array [
            [Function],
            [Function],
          ]
        }
      />
      <VmSizeSelector
        vmSizes={
          Object {
            "amlCompute": Array [],
          }
        }
      />
      <Accordion
        exclusive={true}
        items={
          Array [
            Object {
              "collapsed": true,
              "disabled": false,
              "element": <React.Fragment>
                <minNodeInput />
                <MaxNodeInput
                  minNodeCount="0"
                />
              </React.Fragment>,
              "title": "Additional Settings",
            },
          ]
        }
      />
      <div
        className="form-footer"
      >
        <CustomizedDefaultButton
          onClick={[Function]}
          text="Cancel"
        />
        <CustomizedPrimaryButton
          text="Create"
          type="submit"
        />
      </div>
    </Form>
  </StyledPanelBase>
</Fragment>
`;
